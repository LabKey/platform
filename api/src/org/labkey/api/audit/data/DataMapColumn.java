/*
 * Copyright (c) 2012-2016 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.labkey.api.audit.data;

import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.labkey.api.audit.AbstractAuditTypeProvider;
import org.labkey.api.data.ColumnInfo;
import org.labkey.api.data.DataColumn;
import org.labkey.api.data.DisplayColumn;
import org.labkey.api.data.DisplayColumnFactory;
import org.labkey.api.data.RenderContext;
import org.labkey.api.util.HtmlString;
import org.labkey.api.util.HtmlStringBuilder;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Map;

/**
 * A display column used to render encoded audit datamaps (primarily used to show record changes), in a user readeable format
 */
public class DataMapColumn extends DataColumn
{
    public DataMapColumn(ColumnInfo col)
    {
        super(col);
    }

    @Override @NotNull
    public HtmlString getFormattedHtml(RenderContext ctx)
    {
        HtmlStringBuilder b = HtmlStringBuilder.of();
        HtmlString separator = HtmlString.EMPTY_STRING;
        for (String s : formatColumn(getValue(ctx)))
        {
            b.append(separator);
            separator = HtmlString.BR;
            b.append(s);
        }
        return b.getHtmlString();
    }

    @Override
    public String getTsvFormattedValue(RenderContext ctx)
    {
        return StringUtils.join(formatColumn(getValue(ctx)), "\n");
    }

    @Override
    public Object getDisplayValue(RenderContext ctx)
    {
        return StringUtils.join(formatColumn(getValue(ctx)), "\n");
    }

    @Override
    public boolean isFilterable()
    {
        return false;
    }
    @Override
    public boolean isSortable()
    {
        return false;
    }

    @NotNull
    private List<String> formatColumn(@Nullable Object contents)
    {
        if (contents instanceof String)
        {
            List<String> result = new ArrayList<>();

            for (Map.Entry<String, String> entry : AbstractAuditTypeProvider.decodeFromDataMap((String) contents).entrySet())
            {
                result.add(entry.getKey() + ": " + entry.getValue());
            }
            return result;
        }
        return Collections.emptyList();
    }

    public static class Factory implements DisplayColumnFactory
    {
        @Override
        public DisplayColumn createRenderer(ColumnInfo colInfo)
        {
            return new DataMapColumn(colInfo);
        }
    }
}
