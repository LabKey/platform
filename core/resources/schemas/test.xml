<?xml version="1.0" encoding="UTF-8"?>
<!-- 

 * Copyright (c) 2004-2017 Fred Hutchinson Cancer Research Center
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.

 -->
<tables xsi:schemaLocation="http://labkey.org/data/xml ..\..\..\..\..\schemas\tableInfo.xsd" xmlns="http://labkey.org/data/xml" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
	<table tableName="TestTable" tableDbType="TABLE">
		<columns>
			<column columnName="_ts"/>
			<column columnName="EntityId"/>
			<column columnName="RowId"/>
			<column columnName="CreatedBy"/>
			<column columnName="Created"/>
			<column columnName="Container"/>
			<column columnName="Text"/>
			<column columnName="IntNull"/>
			<column columnName="IntNotNull"/>
			<column columnName="DatetimeNull"/>
			<column columnName="DatetimeNotNull"/>
			<column columnName="RealNull"/>
			<column columnName="BitNull"/>
			<column columnName="BitNotNull"/>
		</columns>
	</table>
    <table tableName="TestTable2" tableDbType="TABLE">
        <columns>
            <column columnName="_ts"/>
            <column columnName="EntityId"/>
            <column columnName="RowId"/>
            <column columnName="CreatedBy"/>
            <column columnName="Created"/>
            <column columnName="Container"/>
            <column columnName="Text"/>
            <column columnName="IntNull"/>
            <column columnName="IntNotNull"/>
            <column columnName="DatetimeNull"/>
            <column columnName="DatetimeNotNull"/>
            <column columnName="RealNull"/>
            <column columnName="BitNull"/>
            <column columnName="BitNotNull"/>
        </columns>
    </table>
    <table tableName="TestTable3" tableDbType="UNKNOWN">
        <description>This is used to test synonyms on SQL Server</description>
        <columns>
            <column columnName="_ts"/>
            <column columnName="EntityId"/>
            <column columnName="RowId"/>
            <column columnName="CreatedBy"/>
            <column columnName="Created"/>
            <column columnName="Container"/>
            <column columnName="Text"/>
            <column columnName="IntNull"/>
            <column columnName="IntNotNull"/>
            <column columnName="DatetimeNull"/>
            <column columnName="DatetimeNotNull"/>
            <column columnName="RealNull"/>
            <column columnName="BitNull"/>
            <column columnName="BitNotNull"/>
        </columns>
    </table>
    <table tableName="Containers2" tableDbType="UNKNOWN">
        <description>This is used to test synonyms on SQL Server</description>
        <columns>
            <column columnName="_ts"/>
            <column columnName="RowId"/>
            <column columnName="EntityId"/>
            <column columnName="CreatedBy"/>
            <column columnName="Created"/>
            <column columnName="Parent"/>
            <column columnName="Name"/>
            <column columnName="SortOrder"/>
            <column columnName="Description"/>
            <column columnName="Type"/>
            <column columnName="Title"/>
            <column columnName="Searchable"/>
        </columns>
    </table>
    <table tableName="ContainerAliases2" tableDbType="UNKNOWN">
        <description>This is used to test synonyms on SQL Server</description>
        <columns>
            <column columnName="Path"/>
            <column columnName="ContainerId"/>
        </columns>
    </table>
    <table tableName="Users2" tableDbType="UNKNOWN">
        <description>This is used to test synonyms on SQL Server</description>
        <columns>
            <column columnName="Email"/>
            <column columnName="_ts"/>
            <column columnName="EntityId"/>
            <column columnName="CreatedBy"/>
            <column columnName="Created"/>
            <column columnName="ModifiedBy"/>
            <column columnName="Modified"/>
            <column columnName="Owner"/>
            <column columnName="UserId"/>
            <column columnName="DisplayName"/>
            <column columnName="FirstName"/>
            <column columnName="LastName"/>
            <column columnName="Phone"/>
            <column columnName="Mobile"/>
            <column columnName="Pager"/>
            <column columnName="IM"/>
            <column columnName="Description"/>
            <column columnName="LastLogin"/>
            <column columnName="Active"/>
            <column columnName="HasPassword"/>
            <column columnName="ExpirationDate"/>
        </columns>
    </table>
</tables>
