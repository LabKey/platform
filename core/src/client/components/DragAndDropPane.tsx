import React, { PureComponent } from 'react';

import { DragDropContext, Draggable, Droppable } from 'react-beautiful-dnd';

import AuthRow from './AuthRow';

interface Props {
    configType: string;
    authConfigs: AuthConfig[];
    providers: AuthConfigProvider[];
    canEdit: boolean;
    isDragDisabled: boolean;
    actions: Actions;
}

export default class DragAndDropPane extends PureComponent<Props> {
    render() {
        const { providers, configType } = this.props;
        const { onDragEnd, ...otherActions } = this.props.actions;

        const DragAndDropAuthRows = this.props.authConfigs.map((authConfig, index) => (
            <Draggable
                key={authConfig.configuration}
                draggableId={authConfig.configuration}
                index={index}
                isDragDisabled={this.props.isDragDisabled}>
                {provided => (
                    <div ref={provided.innerRef} {...provided.draggableProps} {...provided.dragHandleProps}>
                        <AuthRow
                            index={index.toString()}
                            {...authConfig}
                            {...otherActions} // Contains authRow-level functions
                            canEdit={this.props.canEdit}
                            draggable={true}
                            modalType={providers ? providers[authConfig.provider] : null }
                            configType={configType}
                        />
                    </div>
                )}
            </Draggable>
        ));

        return (
            <DragDropContext onDragEnd={onDragEnd}>
                <Droppable droppableId={configType}>
                    {provided => (
                        <div ref={provided.innerRef} {...provided.droppableProps}>
                            {DragAndDropAuthRows}
                            {provided.placeholder}
                        </div>
                    )}
                </Droppable>
            </DragDropContext>
        );
    }
}
