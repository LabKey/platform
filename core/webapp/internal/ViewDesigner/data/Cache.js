/*
 * Copyright (c) 2015-2017 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0
 */
Ext4.define('LABKEY.internal.ViewDesigner.QueryDetailsCache', {

    singleton: true,

    mixins : {
        observable: 'Ext.util.Observable'
    },

    constructor : function() {
        this.cache = {};
        this.results = {};

        this.mixins.observable.constructor.apply(this, arguments);

        this.addEvents('beforecacheresponse');
    },

    callback : function(callback, scope, result) {
        if (Ext4.isFunction(callback)) {
            callback.call(scope, result);
        }
    },

    add : function(keyConfig, result) {
        this.populateCache(this.getKey(keyConfig), result);
    },

    getKey : function(config) {

        if (!Ext4.isString(config.schema) || !Ext4.isString(config.query) ||
                Ext4.isEmpty(config.schema) || Ext4.isEmpty(config.query)) {
            throw 'Invalid cache key. Require a schema and a query.';
        }

        return JSON.stringify({
            schema: config.schema,
            query: config.query,
            view: !Ext4.isEmpty(config.view) ? config.view: undefined,
            fk: Ext4.isString(config.fk) ? config.fk : undefined
        });
    },

    inCache : function(key) {
        return this.cache[key] === true;
    },

    inFlightCache : function(key) {
        return Ext4.isArray(this.cache[key]);
    },

    loadCache : function(key, callback, scope) {
        // The value as an array denotes the cache resource is in flight
        if (!this.cache[key]) {
            this.cache[key] = [];
        }

        if (Ext4.isFunction(callback)) {
            this.cache[key].push({fn: callback, scope: scope});
        }
    },

    populateCache : function(key, result) {

        //console.log('# cols (start):', result.columns.length);
        this.fireEvent('beforecacheresponse', result);
        //console.log('# cols (end)  :', result.columns.length);

        this.results[key] = result;

        var cbs = this.cache[key];
        this.cache[key] = true;

        if (!Ext4.isEmpty(cbs)) {
            Ext4.each(cbs, function(config) {
                this.callback(config.fn, config.scope, result);
            }, this);
        }
    },

    getDetails : function(keyConfig, callback, scope) {
        var key = this.getKey(keyConfig);

        if (this.inCache(key)) {
            // -- cache hit
            this.callback(callback, scope, this.results[key]);
        }
        else if (this.inFlightCache(key)) {
            // -- cache miss, in flight
            this.loadCache(key, callback, scope);
        }
        else {
            // -- cache miss
            this.loadCache(key, callback, scope);

            // Request Details -- assume the current container
            LABKEY.Query.getQueryDetails({
                schemaName: keyConfig.schema,
                queryName: keyConfig.query,
                containerPath: keyConfig.containerPath,
                fk: keyConfig.fk,
                success: function(result) {
                    this.populateCache(key, result);
                },
                scope: this
            });
        }
    }
});

Ext4.define('LABKEY.internal.ViewDesigner.QueryDetailsProxy', {
    extend: 'Ext.data.proxy.Ajax',
    alias: 'proxy.querydetails',
    schema: undefined,
    query: undefined,
    view: undefined,
    url: '',
    read: function(operation, callback, scope) {
        var request = this.buildRequest(operation),
            params = {
                schema: this.schema,
                query: this.query,
                view: this.view,
                containerPath: this.containerPath
            },
            cacheCallback = this.createCacheCallback(request, operation, callback, scope);

        // get the paramters for this request
        if (!Ext4.isEmpty(request.params)) {
            // see if the 'node' parameter is in-use, could be a tree store
            if (!Ext4.isEmpty(request.params.node) &&
                    request.params.node !== LABKEY.internal.ViewDesigner.FieldMetaTreeStore.ROOT_ID) {
                if (operation.node && operation.node.get('fieldKey') !== undefined) {
                    params.fk = operation.node.get('fieldKey');
                }
            }
        }

        LABKEY.internal.ViewDesigner.QueryDetailsCache.getDetails(params, cacheCallback, this);
    },

    createCacheCallback : function(request, operation, callback, scope) {
        var me = this;
        return function(response) {
            me.processResponse(true, operation, request, response, callback, scope);
        }
    }
});