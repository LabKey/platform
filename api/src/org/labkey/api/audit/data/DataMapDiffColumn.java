/*
 * Copyright (c) 2012-2016 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */package org.labkey.api.audit.data;

import org.jetbrains.annotations.NotNull;
import org.labkey.api.audit.AbstractAuditTypeProvider;
import org.labkey.api.data.ColumnInfo;
import org.labkey.api.data.DataColumn;
import org.labkey.api.data.DisplayColumn;
import org.labkey.api.data.DisplayColumnFactory;
import org.labkey.api.data.RenderContext;
import org.labkey.api.data.TableInfo;
import org.labkey.api.query.AliasedColumn;
import org.labkey.api.query.FieldKey;
import org.labkey.api.util.HtmlString;
import org.labkey.api.util.PageFlowUtil;

import java.util.Collections;
import java.util.Map;
import java.util.Set;

/**
 * A column used to help render the changes between two encoded audit datamaps (primarily used to show record changes),
 * in a user readeable format
 */
public class DataMapDiffColumn extends AliasedColumn
{
    public DataMapDiffColumn(TableInfo parent, String name, final ColumnInfo oldValues, final ColumnInfo newValues)
    {
        super(parent, name, oldValues);

        setDisplayColumnFactory(new DisplayColumnFactory()
        {
            @Override
            public DisplayColumn createRenderer(ColumnInfo colInfo)
            {
                return new ChangesDisplayColumn(colInfo, newValues);
            }
        });
    }

    private static class ChangesDisplayColumn extends DataColumn
    {
        ColumnInfo _oldValues;
        ColumnInfo _newValues;

        public ChangesDisplayColumn(ColumnInfo oldValues, ColumnInfo newValues)
        {
            super(oldValues);
            _oldValues = oldValues;
            _newValues = newValues;
        }

        @Override
        public Object getValue(RenderContext ctx)
        {
            Object value = super.getValue(ctx);

            return value != null ? value : "";
        }

        @Override @NotNull
        public HtmlString getFormattedHtml(RenderContext ctx)
        {
            return HtmlString.unsafe(formatColumn(ctx.get(_oldValues.getFieldKey()), ctx.get(_newValues.getFieldKey()), true));
        }

        @Override
        public String getTsvFormattedValue(RenderContext ctx)
        {
            return formatColumn(ctx.get(_oldValues.getFieldKey()), ctx.get(_newValues.getFieldKey()), false);
        }

        @Override
        public Object getDisplayValue(RenderContext ctx)
        {
            return formatColumn(ctx.get(_oldValues.getFieldKey()), ctx.get(_newValues.getFieldKey()), false);
        }

        @Override
        public boolean isFilterable()
        {
            return false;
        }

        @Override
        public boolean isSortable()
        {
            return false;
        }

        @Override
        public void addQueryFieldKeys(Set<FieldKey> keys)
        {
            super.addQueryFieldKeys(keys);

            keys.add(_oldValues.getFieldKey());
            keys.add(_newValues.getFieldKey());
        }

        @NotNull
        private String formatColumn(Object oldContent, Object newContent, boolean html)
        {
            String arrow = html ? "&nbsp;&raquo;&nbsp;" : " > ";
            String lineBreak = html ? "<br/>" : "\n";

            if (oldContent != null || newContent != null)
            {
                String delim = "";
                StringBuilder sb = new StringBuilder();

                Map<String, String> oldValues = Collections.emptyMap();
                Map<String, String> newValues = Collections.emptyMap();

                if (oldContent instanceof String)
                    oldValues = AbstractAuditTypeProvider.decodeFromDataMap((String)oldContent);

                if (newContent instanceof String)
                    newValues = AbstractAuditTypeProvider.decodeFromDataMap((String)newContent);

                for (Map.Entry<String, String> entry : oldValues.entrySet())
                {
                    String oldValue = entry.getValue();
                    if (oldValue == null)
                        oldValue = "";

                    String newValue = newValues.remove(entry.getKey());
                    if (newValue == null)
                        newValue = "";

                    if (!newValue.equals(oldValue))
                    {
                        sb.append(delim);
                        sb.append(html ? PageFlowUtil.filter(entry.getKey()) : entry.getKey());
                        sb.append(": ");
                        sb.append(html ? PageFlowUtil.filter(oldValue) : oldValue);
                        sb.append(arrow);
                        sb.append(html ? PageFlowUtil.filter(newValue) : newValue);

                        delim = lineBreak;
                    }
                }

                for (Map.Entry<String, String> entry : newValues.entrySet())
                {
                    sb.append(delim);
                    sb.append(html ? PageFlowUtil.filter(entry.getKey()) : entry.getKey());
                    sb.append(": ");

                    String newValue = entry.getValue();
                    if (newValue == null)
                        newValue = "";

                    sb.append(arrow);
                    sb.append(html ? PageFlowUtil.filter(newValue) : newValue);

                    delim = lineBreak;
                }

                return sb.toString();
            }
            return "";
        }
    }
}
