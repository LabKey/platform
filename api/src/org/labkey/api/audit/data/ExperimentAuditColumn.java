/*
 * Copyright (c) 2012-2016 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.labkey.api.audit.data;

import org.jetbrains.annotations.Nullable;
import org.labkey.api.data.ColumnInfo;
import org.labkey.api.data.Container;
import org.labkey.api.data.ContainerManager;
import org.labkey.api.data.DataColumn;
import org.labkey.api.data.RenderContext;
import org.labkey.api.exp.api.ExpObject;
import org.labkey.api.util.PageFlowUtil;
import org.labkey.api.util.Pair;
import org.labkey.api.view.ActionURL;

import java.io.IOException;
import java.io.Writer;
import java.util.Set;

/**
 * User: klum
 * Date: Mar 15, 2012
 */
public abstract class ExperimentAuditColumn<ObjectType extends ExpObject> extends DataColumn
{
    protected ColumnInfo _containerId;
    protected ColumnInfo _defaultName;

    public static final String KEY_SEPARATOR = "~~KEYSEP~~";

    public ExperimentAuditColumn(ColumnInfo col, ColumnInfo containerId, ColumnInfo defaultName)
    {
        super(col);
        _containerId = containerId;
        _defaultName = defaultName;
        setTextAlign("left");
    }

    @Override
    public String getName()
    {
        return getColumnInfo().getLabel();
    }

    @Nullable
    protected Container getContainer(RenderContext ctx)
    {
        String cId = (String)ctx.get("ContainerId");
        if (cId == null)
            cId = (String) ctx.get("Container");
        return cId == null ? null : ContainerManager.getForId(cId);
    }

    @Nullable
    protected abstract Pair<ObjectType, ActionURL> getExpValue(RenderContext ctx);

    @Override
    public Object getDisplayValue(RenderContext ctx)
    {
        Pair<ObjectType, ActionURL> value = getExpValue(ctx);
        if (value != null)
        {
            return value.first.getName();
        }

        if (_defaultName != null)
        {
            return extractFromKey3(ctx);
        }
        return null;
    }

    @Override
    public void addQueryColumns(Set<ColumnInfo> columns)
    {
        super.addQueryColumns(columns);
        if (_containerId != null)
            columns.add(_containerId);
        if (_defaultName != null)
            columns.add(_defaultName);
    }

    @Override
    public boolean isFilterable()
    {
        return false;
    }

    @Nullable
    protected abstract String extractFromKey3(RenderContext ctx);

    @Override
    public void renderGridCellContents(RenderContext ctx, Writer out) throws IOException
    {
        Pair<ObjectType, ActionURL> value = getExpValue(ctx);
        if (value != null && value.second != null)
        {
            out.write("<a href=\"" + value.second.getLocalURIString() + "\">" + PageFlowUtil.filter(value.first.getName()) + "</a>");
            return;
        }

        if (_defaultName != null)
        {
            String extracted = extractFromKey3(ctx);
            out.write(extracted != null ? PageFlowUtil.filter(extracted) : "&nbsp;");
        }
        else
            out.write("&nbsp;");
    }

}
