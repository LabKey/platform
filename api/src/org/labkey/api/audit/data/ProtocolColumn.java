/*
 * Copyright (c) 2012-2016 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.labkey.api.audit.data;

import org.jetbrains.annotations.Nullable;
import org.labkey.api.assay.AssayProvider;
import org.labkey.api.assay.AssayService;
import org.labkey.api.assay.AssayUrls;
import org.labkey.api.data.ColumnInfo;
import org.labkey.api.data.Container;
import org.labkey.api.data.RenderContext;
import org.labkey.api.exp.api.ExpProtocol;
import org.labkey.api.exp.api.ExperimentService;
import org.labkey.api.exp.api.ExperimentUrls;
import org.labkey.api.util.PageFlowUtil;
import org.labkey.api.util.Pair;
import org.labkey.api.view.ActionURL;

/**
 * User: klum
 * Date: Mar 15, 2012
 */
public class ProtocolColumn extends ExperimentAuditColumn<ExpProtocol>
{
    public ProtocolColumn(ColumnInfo col, ColumnInfo containerId, @Nullable ColumnInfo defaultName)
    {
        super(col, containerId, defaultName);
    }

    @Nullable
    @Override
    protected Pair<ExpProtocol, ActionURL> getExpValue(RenderContext ctx)
    {
        Object protocolId = getBoundColumn().getValue(ctx);

        if (protocolId != null)
        {
            Container c = getContainer(ctx);
            if (c != null)
            {
                ExpProtocol protocol;

                if (protocolId instanceof Integer)
                    protocol = ExperimentService.get().getExpProtocol((Integer) protocolId);
                else
                    protocol = ExperimentService.get().getExpProtocol(protocolId.toString());

                AssayProvider provider = null;
                if (protocol != null)
                    provider = AssayService.get().getProvider(protocol);

                ActionURL url = null;
                if (provider != null)
                    url = PageFlowUtil.urlProvider(AssayUrls.class).getAssayRunsURL(c, protocol);
                else if (protocol != null)
                    url = PageFlowUtil.urlProvider(ExperimentUrls.class).getProtocolDetailsURL(protocol);
                return protocol == null ? null : new Pair<>(protocol, url);
            }
        }
        return null;
    }

    @Override
    protected String extractFromKey3(RenderContext ctx)
    {
        Object value = _defaultName.getValue(ctx);
        if (value == null)
            return null;
        String[] parts = value.toString().split(KEY_SEPARATOR);
        if (parts.length != 2)
            return null;
        return parts[0];
    }

}
