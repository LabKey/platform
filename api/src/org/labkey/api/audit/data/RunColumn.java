/*
 * Copyright (c) 2012-2016 LabKey Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.labkey.api.audit.data;

import org.jetbrains.annotations.Nullable;
import org.labkey.api.data.ColumnInfo;
import org.labkey.api.data.Container;
import org.labkey.api.data.RenderContext;
import org.labkey.api.exp.api.ExpProtocol;
import org.labkey.api.exp.api.ExpRun;
import org.labkey.api.exp.api.ExperimentService;
import org.labkey.api.exp.api.ExperimentUrls;
import org.labkey.api.assay.AssayProvider;
import org.labkey.api.assay.AssayService;
import org.labkey.api.assay.AssayUrls;
import org.labkey.api.util.PageFlowUtil;
import org.labkey.api.util.Pair;
import org.labkey.api.view.ActionURL;

/**
 * User: klum
 * Date: Mar 15, 2012
 */
public class RunColumn extends ExperimentAuditColumn<ExpRun>
{
    public RunColumn(ColumnInfo col, ColumnInfo containerId, @Nullable ColumnInfo defaultName)
    {
        super(col, containerId, defaultName);
    }

    protected String extractFromKey3(RenderContext ctx)
    {
        Object value = _defaultName.getValue(ctx);
        if (value == null)
            return null;
        String[] parts = value.toString().split(KEY_SEPARATOR);
        if (parts.length != 2)
            return null;
        return parts[1].length() > 0 ? parts[1] : null;
    }

    @Override
    @Nullable
    protected Pair<ExpRun, ActionURL> getExpValue(RenderContext ctx)
    {
        String runLsid = (String) getBoundColumn().getValue(ctx);
        if (runLsid != null)
        {
            Container c = getContainer(ctx);
            if (c != null)
            {
                ExpRun run = ExperimentService.get().getExpRun(runLsid);
                ExpProtocol protocol = null;
                if (run != null)
                    protocol = run.getProtocol();
                AssayProvider provider = null;
                if (protocol != null)
                    provider = AssayService.get().getProvider(protocol);

                ActionURL url = null;
                if (provider != null)
                    url = PageFlowUtil.urlProvider(AssayUrls.class).getAssayResultsURL(c, protocol, run.getRowId());
                else if (run != null)
                    url = PageFlowUtil.urlProvider(ExperimentUrls.class).getRunGraphURL(run);

                return run == null ? null : new Pair<>(run, url);
            }
        }
        return null;
    }
}
