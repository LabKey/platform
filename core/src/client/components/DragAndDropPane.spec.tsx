import React from 'react';

import { shallow } from 'enzyme';
import EnzymeToJson from 'enzyme-to-json';

import { SSO_CONFIGURATIONS, PRIMARY_PROVIDERS, PROVIDERS } from '../../../test/data';

import DragAndDropPane from './DragAndDropPane';

describe('<DragAndDropPane/>', () => {
    let component;

    beforeEach(() => {
        const onDragEnd = (result: {[key:string]: any}): void => {};
        const toggleModalOpen = (modalOpen: boolean): void => {};
        const onDelete = (configuration: number, configType: string): void => {};
        const updateAuthRowsAfterSave = (config: string, configType: string): void => {};
        const actionFns = {
            onDragEnd: onDragEnd, toggleModalOpen: toggleModalOpen,
            onDelete: onDelete, updateAuthRowsAfterSave: updateAuthRowsAfterSave
        };

        component = (
            <DragAndDropPane
                configType={"ssoConfigurations"}
                authConfigs={SSO_CONFIGURATIONS}
                providers={PROVIDERS}
                canEdit={true}
                isDragDisabled={false}
                actions={actionFns}
            />
        );
    });

    test('Drag is disabled', () => {
        const wrapper = shallow<DragAndDropPane>(component);

        wrapper.setProps({ isDragDisabled: true });
        expect(EnzymeToJson(wrapper)).toMatchSnapshot();
    });

    test('View-only', () => {
        const wrapper = shallow<DragAndDropPane>(component);
        wrapper.setProps({ canEdit: false });

        expect(EnzymeToJson(wrapper)).toMatchSnapshot();
    });

    test('Editable', () => {
        const wrapper = shallow(component);
        expect(EnzymeToJson(wrapper)).toMatchSnapshot();
    });
});
