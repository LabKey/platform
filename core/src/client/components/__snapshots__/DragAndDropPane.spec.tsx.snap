// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DragAndDropPane/> Drag is disabled 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <Connect(Droppable)
    direction="vertical"
    droppableId="ssoConfigurations"
    ignoreContainerClipping={false}
    isCombineEnabled={false}
    isDropDisabled={false}
    type="DEFAULT"
  >
    <Component />
  </Connect(Droppable)>
</DragDropContext>
`;

exports[`<DragAndDropPane/> Editable 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <Connect(Droppable)
    direction="vertical"
    droppableId="ssoConfigurations"
    ignoreContainerClipping={false}
    isCombineEnabled={false}
    isDropDisabled={false}
    type="DEFAULT"
  >
    <Component />
  </Connect(Droppable)>
</DragDropContext>
`;

exports[`<DragAndDropPane/> View-only 1`] = `
<DragDropContext
  onDragEnd={[Function]}
>
  <Connect(Droppable)
    direction="vertical"
    droppableId="ssoConfigurations"
    ignoreContainerClipping={false}
    isCombineEnabled={false}
    isDropDisabled={false}
    type="DEFAULT"
  >
    <Component />
  </Connect(Droppable)>
</DragDropContext>
`;
